packages/realm-server/billing/billing-queries.ts

```diff
@@ -0,0 +1,327 @@
+  dbAdapter: DBAdapter,
+  stripeCustomerId: string,
+): Promise<User | null> {
+  let results = await query(dbAdapter, [
+    `SELECT * FROM users WHERE stripe_customer_id = `,
+    param(stripeCustomerId),
+  ]);
+
+  if (results.length !== 1) {
+    return null;
+  }
+
+  return {
+    id: results[0].id,
+    matrixUserId: results[0].matrix_user_id,
+    stripeCustomerId: results[0].stripe_customer_id,
+  } as User;
+}
+
+export async function insertSubscriptionCycle(
+  dbAdapter: DBAdapter,
+  subscriptionCycle: {
+    subscriptionId: string;
+    periodStart: number;
+    periodEnd: number;
+  },
+): Promise<SubscriptionCycle> {
+  let { valueExpressions, nameExpressions } = asExpressions({
+    subscription_id: subscriptionCycle.subscriptionId,
+    period_start: subscriptionCycle.periodStart,
+    period_end: subscriptionCycle.periodEnd,
+  });
+
+  let result = await query(dbAdapter, [
+    `INSERT INTO subscription_cycles`,
+    ...addExplicitParens(separatedByCommas(nameExpressions)),
+    ` VALUES`,
+    ...addExplicitParens(separatedByCommas(valueExpressions)),
+    ` RETURNING *`,
+  ] as Expression);
+
+  return {
+    id: result[0].id,
+    subscriptionId: result[0].subscription_id,
+    periodStart: result[0].period_start,
+    periodEnd: result[0].period_end,
+  } as SubscriptionCycle;
+}
+
+export async function insertSubscription(
+  dbAdapter: DBAdapter,
+  subscription: {
+    user_id: string;
+    plan_id: string;
+    started_at: number;
+    status: string;
+    stripe_subscription_id: string;
+  },
+): Promise<Subscription> {
+  let { valueExpressions, nameExpressions } = asExpressions({
+    user_id: subscription.user_id,
+    plan_id: subscription.plan_id,
+    started_at: subscription.started_at,
+    status: subscription.status,
+    stripe_subscription_id: subscription.stripe_subscription_id,
+  });
+
+  let result = await query(dbAdapter, [
+    `INSERT INTO subscriptions`,
+    ...addExplicitParens(separatedByCommas(nameExpressions)),
+    ` VALUES`,
+    ...addExplicitParens(separatedByCommas(valueExpressions)),
+    ` RETURNING *`,
+  ] as Expression);
+
+  return {
+    id: result[0].id,
+    userId: result[0].user_id,
+    planId: result[0].plan_id,
+    startedAt: result[0].started_at,
+    status: result[0].status,
+    stripeSubscriptionId: result[0].stripe_subscription_id,
+  } as Subscription;
+}
+
+export async function addToCreditsLedger(
+  dbAdapter: DBAdapter,
+  ledgerEntry: Omit<LedgerEntry, 'id'>,
+) {
+  let { valueExpressions, nameExpressions } = asExpressions({
+    user_id: ledgerEntry.userId,
+    credit_amount: ledgerEntry.creditAmount,
+    credit_type: ledgerEntry.creditType,
+    subscription_cycle_id: ledgerEntry.subscriptionCycleId,
+  });
+
+  await query(dbAdapter, [
+    `INSERT INTO credits_ledger`,
+    ...addExplicitParens(separatedByCommas(nameExpressions)),
+    ` VALUES`,
+    ...addExplicitParens(separatedByCommas(valueExpressions)),
+  ] as Expression);
+}
```

Comment:
in the expressions.ts module we have a function `upsert` that just takes in nameExpressions and valueExpressions so that the calls can be very simple, like:

``` ts   
await this.query([
      ...upsert(
        'boxel_index',
        'boxel_index_pkey',
        nameExpressions,
        valueExpressions,
      ),
    ]);
```

let's make a function for `insert` that also takes in a `nameExpression` and a `valueExpression`, so that we can just write:
```ts
await query(insert('table_name', nameExpressions, valueExpressions))
```

Reply:
Seems like a good idea, I'll do it

Reply:
Done https://github.com/cardstack/boxel/pull/1719/commits/9e78818a27854a49bd0a221dc3c1dfa005014b81